symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -190,27 +190,37 @@ instance
   ( Arithmetic a, Binary a, Representable i, Binary (Rep i), Ord (Rep i)
   , o ~ U1) => MonadCircuit (Var a i) a (WitnessF (Var a i) a) (State (ArithmeticCircuit a i o)) where
 
-    unconstrained wf = do
-      let v = toVar @a wf
-      -- TODO: forbid reassignment of variables
-      zoom #acWitness . modify $ insert v $ \i w -> witnessF wf $ \case
-        SysVar (InVar inV) -> index i inV
-        SysVar (NewVar newV) -> w ! newV
-        ConstVar cV -> fromConstant cV
-      return (SysVar (NewVar v))
+    unconstrained wf =
+      let res = witnessF wf $ \case
+            ConstVar cV -> Just cV
+            _           -> Nothing
+      in case res of
+        Just cV -> return (ConstVar cV)
+        Nothing -> do
+          let v = toVar @a wf
+          zoom #acWitness . modify $ insert v $ \i w -> witnessF wf $ \case
+            SysVar (InVar inV) -> index i inV
+            SysVar (NewVar newV) -> w ! newV
+            ConstVar cV -> fromConstant cV
+          return $ SysVar (NewVar v)
+
 
     constraint p =
       let
         evalConstVar = \case
           SysVar sysV -> var sysV
-          ConstVar cV -> fromConstant cV
+          ConstVar cV -> if cV == zero
```

Comment:
There's nothing wrong with nonzero constants inside constraint; I think you meant to check the result of `p evalConstVar` if it is zero